### PR TITLE
[infra] Use "-fsanitize=fuzzer-no-link" instead of coverage flags.

### PR DESCRIPTION
**Pull request by Dor1s**
Tested with tor project, seems working fine. I'll probably test more with something else and then land it next week when be able to keep a closer eye on all of the builds.

### CODE REVIEW

File: infra/base-images/base-builder/Dockerfile [hunk 1]
```diff
@@ -27,7 +27,7 @@ ENV SANITIZER_FLAGS_memory "-fsanitize=memory -fsanitize-memory-track-origins"
 ENV SANITIZER_FLAGS_profile "-fprofile-instr-generate -fcoverage-mapping -pthread"
 
 # Default build flags for coverage.
-ENV COVERAGE_FLAGS="-fsanitize-coverage=trace-pc-guard,trace-cmp"
+ENV COVERAGE_FLAGS="-fsanitize=fuzzer-no-link"
 
 # Coverage flags for generating coverage reports.
```
Comment (Dor1s) [inline infra/base-images/base-builder/Dockerfile]: I guess we don't want to change that, do we?

Comment (kcc) [reply infra/base-images/base-builder/Dockerfile]: At least not now.

File: infra/base-images/base-builder/Dockerfile [hunk 2]
```diff
@@ -27,7 +27,7 @@ ENV SANITIZER_FLAGS_memory "-fsanitize=memory -fsanitize-memory-track-origins"
 ENV SANITIZER_FLAGS_profile "-fprofile-instr-generate -fcoverage-mapping -pthread"
 
 # Default build flags for coverage.
-ENV COVERAGE_FLAGS="-fsanitize-coverage=trace-pc-guard,trace-cmp"
+ENV COVERAGE_FLAGS="-fsanitize=fuzzer-no-link"
```
Comment (kcc) [inline infra/base-images/base-builder/Dockerfile]: Thanks! 
The change itself is perfect, but may need much more testing. 
This small change actually brings a big difference.

**Comment by Dor1s** [comment]
Testing this with libreoffice:
### Old build
```
root@ba7c25729cba:/out/cf# ./lwpfuzzer 
INFO: Seed: 1785918414
INFO: Loaded 1 modules   (814029 guards): 814029 [0x91100e0, 0x942b014), 
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
#2	INITED cov: 6315 ft: 154 corp: 1/1b exec/s: 0 rss: 196Mb
#8	NEW    cov: 6331 ft: 169 corp: 2/21b exec/s: 0 rss: 196Mb L: 20/20 MS: 1 InsertRepeatedBytes-
#9	NEW    cov: 6339 ft: 190 corp: 3/151b exec/s: 0 rss: 196Mb L: 130/130 MS: 2 InsertRepeatedBytes-InsertRepeatedBytes-
#10	NEW    cov: 6339 ft: 192 corp: 4/365b exec/s: 0 rss: 196Mb L: 214/214 MS: 3 InsertRepeatedBytes-InsertRepeatedBytes-InsertRepeatedBytes-
<...>
#115796	REDUCE cov: 7062 ft: 2190 corp: 174/72Kb exec/s: 3216 rss: 699Mb L: 125/4096 MS: 2 InsertByte-EraseBytes-

```

### New build
```
root@ba7c25729cba:/out# ./lwpfuzzer 
INFO: Seed: 1532852106
INFO: Loaded 1 modules   (814037 inline 8-bit counters): 814037 [0x8c7a120, 0x8d40cf5), 
INFO: Loaded 1 PC tables (814037 PCs): 814037 [0x8d40cf8,0x99aca48), 
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
#2	INITED cov: 171 ft: 155 corp: 1/1b exec/s: 0 rss: 163Mb
	NEW_FUNC[0/7]: 0x542660 in SvStream::GetError() const /src/libreoffice/include/tools/stream.hxx:205
	NEW_FUNC[1/7]: 0x543df0 in ErrCode::IgnoreWarning() const /src/libreoffice/include/vcl/errcode.hxx:89
#16	NEW    cov: 190 ft: 184 corp: 2/61b exec/s: 0 rss: 165Mb L: 60/60 MS: 4 CopyPart-InsertRepeatedBytes-ChangeBinInt-InsertRepeatedBytes-
<...>
#115792	REDUCE cov: 886 ft: 2129 corp: 186/107Kb exec/s: 904 rss: 833Mb L: 171/4096 MS: 3 ChangeBinInt-ChangeBinInt-EraseBytes-

```

Execution speed and `cov:` values look surprising. Re-running old build with `seed` value used for running the new build:
```
root@ba7c25729cba:/out/cf# ./lwpfuzzer -seed=1532852106 -runs=115800
INFO: Seed: 1532852106
INFO: Loaded 1 modules   (814029 guards): 814029 [0x91100e0, 0x942b014), 
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
#2	INITED cov: 6321 ft: 154 corp: 1/1b exec/s: 0 rss: 130Mb
#22	NEW    cov: 6345 ft: 182 corp: 2/45b exec/s: 0 rss: 130Mb L: 44/44 MS: 5 CopyPart-CMP-EraseBytes-InsertByte-InsertRepeatedBytes- DE: "soffice"-

<...>
#115799	REDUCE cov: 6946 ft: 2325 corp: 211/209Kb exec/s: 654 rss: 505Mb L: 2744/3978 MS: 5 ShuffleBytes-PersAutoDict-PersAutoDict-ChangeBinInt-EraseBytes- DE: "\x03\x00\x00\x00\x00\x00\x00\x00"-"\x03\x00\x00\x00\x00\x00\x00\x00"-
```

Ok, speed is even lower now (i.e. old build is slower than new one), so the change doesn't seem to decrease it. To double check, run the new build with `seed` value used for running the old build:
```
root@ba7c25729cba:/out# ./lwpfuzzer -seed=1785918414 -runs=115800
INFO: Seed: 1785918414
INFO: Loaded 1 modules   (814037 inline 8-bit counters): 814037 [0x8c7a120, 0x8d40cf5), 
INFO: Loaded 1 PC tables (814037 PCs): 814037 [0x8d40cf8,0x99aca48), 
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
#2	INITED cov: 171 ft: 155 corp: 1/1b exec/s: 0 rss: 157Mb
	NEW_FUNC[0/4]: 0x542660 in SvStream::GetError() const /src/libreoffice/include/tools/stream.hxx:205
	NEW_FUNC[1/4]: 0x543df0 in ErrCode::IgnoreWarning() const /src/libreoffice/include/vcl/errcode.hxx:89
#8	NEW    cov: 182 ft: 170 corp: 2/21b exec/s: 0 rss: 158Mb L: 20/20 MS: 1 InsertRepeatedBytes-
	NEW_FUNC[0/3]: 0x5524f0 in OpenStormBento::LtcBenContainer::SeekToPosition(unsigned long) /src/libreoffice/lotuswordpro/source/filter/bencont.cxx:201
	NEW_FUNC[1/3]: 0x62f970 in OpenStormBento::CBenTOCReader::SearchForLabel(unsigned char*) /src/libreoffice/lotuswordpro/source/filter/tocread.cxx:146
#9	NEW    cov: 190 ft: 192 corp: 3/151b exec/s: 0 rss: 158Mb L: 130/130 MS: 2 InsertRepeatedBytes-InsertRepeatedBytes-
<...>
#115800	DONE   cov: 887 ft: 1940 corp: 169/77Kb exec/s: 2184 rss: 806Mb
```

In that case new build looks slower than the old one, but we may assume that it's relatively the same as there is not too much date to evaluate.

## Coverage
Run both old and new build over the existing corpus:

### Old build
```
root@ba7c25729cba:/out/cf# ./lwpfuzzer -runs=0 ../corpus
INFO: Seed: 2481918259
INFO: Loaded 1 modules   (814029 guards): 814029 [0x91100e0, 0x942b014), 
INFO:     4345 files found in ../corpus
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 990456 bytes
INFO: seed corpus: files: 4345 min: 8b max: 990456b total: 4656064b rss: 130Mb
#512	pulse  cov: 7795 ft: 4617 corp: 507/19Kb exec/s: 64 rss: 626Mb
#1024	pulse  cov: 10241 ft: 8714 corp: 1010/60Kb exec/s: 37 rss: 759Mb
#2048	pulse  cov: 10691 ft: 13744 corp: 2032/216Kb exec/s: 62 rss: 759Mb
#4096	pulse  cov: 13653 ft: 27532 corp: 4072/1650Kb exec/s: 75 rss: 799Mb
#4351	INITED cov: 14554 ft: 31822 corp: 4328/4540Kb exec/s: 76 rss: 799Mb
#4351	DONE   cov: 14554 ft: 31822 corp: 4328/4540Kb exec/s: 76 rss: 799Mb
Done 4351 runs in 57 second(s)
```

### New build
```
root@ba7c25729cba:/out# ./lwpfuzzer -runs=0 ./corpus
INFO: Seed: 2400718029
INFO: Loaded 1 modules   (814037 inline 8-bit counters): 814037 [0x8c7a120, 0x8d40cf5), 
INFO: Loaded 1 PC tables (814037 PCs): 814037 [0x8d40cf8,0x99aca48), 
INFO:     4345 files found in ./corpus
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 990456 bytes
INFO: seed corpus: files: 4345 min: 8b max: 990456b total: 4656064b rss: 427Mb
#512	pulse  cov: 1792 ft: 4625 corp: 507/19Kb exec/s: 73 rss: 655Mb
#1024	pulse  cov: 4313 ft: 8750 corp: 1013/61Kb exec/s: 44 rss: 808Mb
#2048	pulse  cov: 4785 ft: 13794 corp: 2035/217Kb exec/s: 70 rss: 808Mb
#4096	pulse  cov: 8512 ft: 27630 corp: 4074/1650Kb exec/s: 85 rss: 876Mb
#4351	INITED cov: 9415 ft: 31925 corp: 4329/4537Kb exec/s: 85 rss: 876Mb
#4351	DONE   cov: 9415 ft: 31925 corp: 4329/4537Kb exec/s: 85 rss: 876Mb
Done 4351 runs in 51 second(s)
```

The new build looks slightly faster, but `cov:` value is significantly lower than it used to be, i.e. `14554` vs `9415`.

@kcc, is that expected?

**Comment by kcc** [comment]
Yes, this is expected. 
The new code has different logic for accounting the code that happens at initialization time (i.e. once). 

What's more important is that the values for `ft: ` are similar (`ft:`, or features is what libFuzzer actually looks at)

**Comment by Dor1s** [comment]
ffmpeg build works as well (if I use working revision of x265, since current ToT is broken https://github.com/google/oss-fuzz/issues/981)

## A bit of fuzzing with a fixed seed value

### Old build
```
root@5b38b7f8fbc7:/out/cf# ./ffmpeg_AV_CODEC_ID_HEVC_fuzzer -seed=123456789 -runs=100000
INFO: Seed: 123456789
INFO: Loaded 1 modules   (26210 guards): 26210 [0xd87be8, 0xda1570), 
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
#2  INITED cov: 364 ft: 344 corp: 1/1b exec/s: 0 rss: 42Mb
#16 NEW    cov: 386 ft: 389 corp: 2/47b exec/s: 0 rss: 43Mb L: 46/46 MS: 4 ShuffleBytes-ChangeBit-ChangeByte-InsertRepeatedBytes-
<...>
#100000 DONE   cov: 523 ft: 1072 corp: 191/164Kb exec/s: 5882 rss: 444Mb
```

### New build
```
root@5b38b7f8fbc7:/out# ./ffmpeg_AV_CODEC_ID_HEVC_fuzzer -seed=123456789 -runs=100000
INFO: Seed: 123456789
INFO: Loaded 1 modules   (26250 inline 8-bit counters): 26250 [0xddfd48, 0xde63d2), 
INFO: Loaded 1 PC tables (26250 PCs): 26250 [0xa99310,0xaffbb0), 
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 4096 bytes
INFO: A corpus is not provided, starting from an empty corpus
#2  INITED cov: 344 ft: 345 corp: 1/1b exec/s: 0 rss: 43Mb
  NEW_FUNC[0/4]: 0x67e0e0 in find_next_start_code /src/ffmpeg/libavcodec/h2645_parse.c:249
  NEW_FUNC[1/4]: 0x86ce70 in av_log_default_callback /src/ffmpeg/libavutil/log.c:303
#16 NEW    cov: 366 ft: 390 corp: 2/47b exec/s: 0 rss: 45Mb L: 46/46 MS: 4 ShuffleBytes-ChangeBit-ChangeByte-InsertRepeatedBytes-

<...>
#100000 DONE   cov: 533 ft: 1212 corp: 232/248Kb exec/s: 6250 rss: 448Mb
```
Looks quite similar and legit.

## Running over the existing corpus

### Old build
```
root@5b38b7f8fbc7:/out/cf# ./ffmpeg_AV_CODEC_ID_HEVC_fuzzer -runs=0 -timeout=5 -rss_limit_mb=4096 ../corpus
INFO: Seed: 3429512016
INFO: Loaded 1 modules   (26210 guards): 26210 [0xd87be8, 0xda1570), 
INFO:     4879 files found in ../corpus
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 1048576 bytes
INFO: seed corpus: files: 4879 min: 4b max: 1610303b total: 1458891846b rss: 44Mb
#256  pulse  cov: 4038 ft: 12473 corp: 254/57Kb exec/s: 128 rss: 199Mb
#512  pulse  cov: 4985 ft: 20213 corp: 510/189Kb exec/s: 85 rss: 377Mb
#1024 pulse  cov: 5898 ft: 29263 corp: 1021/804Kb exec/s: 56 rss: 408Mb
#2048 pulse  cov: 6390 ft: 37546 corp: 2045/12706Kb exec/s: 8 rss: 549Mb
#4096 pulse  cov: 7886 ft: 51433 corp: 4093/687Mb exec/s: 5 rss: 1376Mb
#4880 INITED cov: 7906 ft: 53243 corp: 4876/1388Mb exec/s: 4 rss: 2148Mb
#4880 DONE   cov: 7906 ft: 53243 corp: 4876/1388Mb exec/s: 4 rss: 2148Mb
Done 4880 runs in 1095 second(s)
```

### New build

```
root@5b38b7f8fbc7:/out# ./ffmpeg_AV_CODEC_ID_HEVC_fuzzer -runs=0 -timeout=5 -rss_limit_mb=4096  ./corpus
INFO: Seed: 2025666828
INFO: Loaded 1 modules   (26250 inline 8-bit counters): 26250 [0xddfd48, 0xde63d2), 
INFO: Loaded 1 PC tables (26250 PCs): 26250 [0xa99310,0xaffbb0), 
INFO:     4879 files found in ./corpus
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 1048576 bytes
INFO: seed corpus: files: 4879 min: 4b max: 1610303b total: 1458891846b rss: 45Mb
#256  pulse  cov: 4015 ft: 12505 corp: 254/57Kb exec/s: 128 rss: 200Mb
#512  pulse  cov: 4965 ft: 20269 corp: 510/189Kb exec/s: 102 rss: 378Mb
#1024 pulse  cov: 5878 ft: 29320 corp: 1022/806Kb exec/s: 68 rss: 418Mb
#2048 pulse  cov: 6355 ft: 37636 corp: 2045/12706Kb exec/s: 9 rss: 560Mb
#4096 pulse  cov: 7867 ft: 51541 corp: 4092/687Mb exec/s: 6 rss: 1376Mb
#4880 INITED cov: 7887 ft: 53356 corp: 4875/1388Mb exec/s: 4 rss: 2171Mb
#4880 DONE   cov: 7887 ft: 53356 corp: 4875/1388Mb exec/s: 4 rss: 2171Mb
Done 4880 runs in 1013 second(s)
```

Same here: the new build is relatively faster, has a little bit more features reported, looks good.

I'll land this tomorrow morning to be able to keep an eye on the other builds that may start failing.

**Comment by Dor1s** [comment]
I don't see any breakages related to that change.

**Pull request merged**